allennlp/training/trainer.py

```diff
@@ -201,16 +378,32 @@ def _train_epoch(self, epoch: int) -> dict:
         # Get tqdm for the training batches
         train_generator = self._iterator(self._train_data,
                                          num_epochs=1,
-                                         cuda_device=self._cuda_device)
+                                         cuda_device=self._iterator_device)
         num_training_batches = self._iterator.get_num_batches(self._train_data)
         train_generator_tqdm = Tqdm.tqdm(train_generator,
                                          total=num_training_batches)
         self._last_log = time.time()
+        last_save_time = time.time()
+
         batch_num = 0
+        if self._batch_num_total is None:
+            self._batch_num_total = 0
+
+        if self._histogram_interval is not None:
+            if hasattr(self._model, 'histogram_parameters'):
```

Comment:
For third party users, requiring them to add an empty method for an option they may not use seems pretty heavy weight.  For example, I'm using this `Trainer` without subclasses the allennlp `Model`, but by implementing the interface.  There are already a number of stub like functions that I don't use but are necessary to get it to work.  An optional attribute is easier to use since it only requires the user to know about it if they need it, versus always having to know about it whether they use it or not.

Reply:
Can't this logic live below, where `histogram_parameters` is actually used?

Also, a method on `Model` that defaults to your `else` case would simplify this logic and not require any user to know about it unless they used it, just like your attribute.  But, it's more type-safe and it makes the library code easier to understand and maintain.  I'm a strong vote in favor of the method instead of `hasattr` duck typing.

Reply:
I'm not sure where you mean to add it below?  I moved it above the loop so we aren't running this every time we decide to send histograms to tensorboard.

Reply:
The largest drawback to adding the method to the `Model` interface is for users who don't want to subclass the allennlp `Model` for whatever reason (e.g., me) still need to provide the stub.  It depends on whether you want the abstractions in allennlp to be heavyweight or lightweight.  Ideally one should be able to just swap in a generic `torch.nn.Module` that returns a loss and train that (and this isn't the case now, it's still necessary to add a few empty stubs).  Adding the method to the required interface makes it harder to use just the `Trainer` without the rest of allennlp.

Reply:
Because of all of the comments and the collapsed blocks, I missed that it was used inside a loop - feel free to ignore my comment about that.

Why don't you subclass `Model`?  You're presumably subclassing `Module` already - what's stopping you from changing that to `Model`?

And ok, you're right that using a method instead of `hasattr` makes it harder to use a `Trainer` without a `Model`.  But doing it your way favors that use case _at the expense_ of the maintainability of our library.  I've seen Keras's code, that has `hasattrs` sprinkled everywhere, and it's a mess to understand what's going on.  I don't think we should go in that direction.  I think changing `class MyModule(torch.nn.Module)` to `class MyModule(allennlp.models.Model)` is extremely lightweight - it's a one-line change.  I don't _think_ that anything else is required.  Am I missing something?

Reply:
For starters, subclassing requires calling `super(..., self).__init__(...)`.  The allennlp `Model` requires an input `Vocabulary` which I'm not using.  So subclassing pulls more allennlp specific things, or the subclass would have to call  `super(MyModule, self).__init__(None)` then hope allennlp doesn't change the signature of `__init__` or actually do something in the constructor that requires the `Vocabulary` to be valid.

Reply:
I'm with Matt G on this one i'm afraid - I find it tolerable to do the type checking stuff when you're checking basic python data structures, like in the histogram logging, but for code that we've written, it's much less obvious when or why you should have attributes. E.g if I add an attribute to a class, I don't expect it to change it's behaviour externally, apart from in very special cases like the histogram logging.

Reply:
Great, that's a good point.  @DeNeutoy, @joelgrus, any reason we shouldn't make the vocab an optional input to `Model`?  It makes it so that `forward_on_instances` will fail, so we could log a warning or something, but it would make things easier for @matt-peters.  Matt, anything else?

Reply:
And just to be clear: if the blocking issues for using `Model` are easily solvable, I would _strongly_ prefer fixing them to making changes to `Trainer` that try to handle things that aren't `Models` - it just makes it too hard to maintain `Trainer`.

Reply:
Nothing else except what I've said previously: subclassing provides significant cognitive drain for a user.  It requires them to understand the lower level details of the library they are subclassing, and makes their code more brittle since they now also rely on allennlp's interface not to change.  From the discussion it seems like we are OK with that.

Reply:
Well, we rely on it to call `forward_on_instance` - but I guess we could just make it optional and raise errors if you try to use any of the methods without overriding them.

Reply:
Except you're _already_ subclassing `Module`.  I'm comfortable making a promise to users that changing a `Module` into a `Model` will always be a one-line change (`Module` already requires you to call `super().__init__()`), with the user not having to worry about anything else unless they want to tweak behavior.

Reply:
Well, thinking about it a bit more, that promise is pretty meaningless, because we also require you to return a dictionary with a `loss` key.  I'm really not sure what additional cognitive drain you get from making that one-line change after you've already figured out our `forward` requirements...

Reply:
I moved the attribute to a `get_parameters_for_histogram_tensorboard_logging` method on `Model`.